executor/src/witgen/identity_processor.rs

```diff
@@ -5,25 +5,75 @@ use itertools::{Either, Itertools};
 use lazy_static::lazy_static;
 use number::FieldElement;
 
+use crate::witgen::machines::Machine;
+
 use super::{
     affine_expression::AffineExpression,
-    machines::{FixedLookup, KnownMachine, Machine},
+    machines::{FixedLookup, KnownMachine},
     rows::RowPair,
     EvalResult, EvalValue, FixedData, IncompleteCause,
 };
 
+/// A list of mutable references to machines.
+pub struct Machines<'a, 'b, T: FieldElement> {
+    machines: Vec<&'b mut KnownMachine<'a, T>>,
+}
+
+impl<'a, 'b, T: FieldElement> Machines<'a, 'b, T> {
+    /// Applies a given function to the ith machine and all other machines.
+    /// With #515 implemented, this could be removed as there would not be a need to split the
+    /// list of machines.
+    pub fn split<'c, R>(
+        &'c mut self,
+        i: usize,
+        f: impl FnOnce(&'c mut KnownMachine<'a, T>, Machines<'a, 'c, T>) -> R,
+    ) -> R {
+        let (before, after) = self.machines.split_at_mut(i);
+        let (current, after) = after.split_at_mut(1);
+        let current: &'c mut KnownMachine<'a, T> = current.first_mut().unwrap();
+
+        // Reborrow machines to convert from `&'c mut &'b mut KnownMachine<'a, T>` to
+        // `&'c mut KnownMachine<'a, T>`.
+        let others: Vec<&'c mut KnownMachine<'a, T>> = before
+            .iter_mut()
+            .chain(after.iter_mut())
+            .map(|m| &mut **m)
+            .collect();
+
+        f(current, others.into())
+    }
+
+    pub fn len(&self) -> usize {
+        self.machines.len()
+    }
+}
+
+impl<'a, 'b, T: FieldElement> From<Vec<&'b mut KnownMachine<'a, T>>> for Machines<'a, 'b, T> {
+    fn from(machines: Vec<&'b mut KnownMachine<'a, T>>) -> Self {
+        Self { machines }
+    }
+}
+
+impl<'a, 'b, T: FieldElement> From<&'b mut Vec<KnownMachine<'a, T>>> for Machines<'a, 'b, T> {
+    fn from(machines: &'b mut Vec<KnownMachine<'a, T>>) -> Self {
```

Comment:
Is it possible to generalize this from a vector to arbitrary collections? Not a blocker though.